File: iconics-java/src/main/java/com/open/iconics/IconicsBrush.java
package com.open.iconics;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Paint;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.open.iconics.animation.IconicsAnimationProcessor;

/**
 * Helper class to control applying state changes to colors and paint. Also for compact providing
 * into {@link IconicsAnimationProcessor AnimationProcessor}
 */
public class IconicsBrush<T extends Paint> {
    @Nullable
    private ColorStateList mColors;
    @NonNull
    private final T mPaint;
    @Nullable
    private int[] mState;

    public IconicsBrush(@NonNull T paint) {
        mPaint = paint;
    }

    /**
     * @param colors which will be applied on {@link #getPaint()} for drawing current state
     */
    public IconicsBrush<T> setColors(@Nullable ColorStateList colors) {
        mColors = colors;
        return this;
    }

    /**
     * @return colors which applied on {@link #getPaint()} for drawing current state
     */
    @Nullable
    public ColorStateList getColorsList() {
        return mColors;
    }

    /**
     * @return paint. Will be used for drawing something (icon, background etc.)
     */
    @NonNull
    public T getPaint() {
        return mPaint;
    }

    /**
     * @param alpha channel for colors
     */
    public void setAlpha(@IntRange(from = 0, to = 255) int alpha) {
        if (mPaint.getAlpha() != alpha) {
            mPaint.setAlpha(alpha);
        }
    }

    /**
     * @return alpha channel for colors
     */
    @IntRange(from = 0, to = 255)
    public int getAlpha() {
        return mPaint.getAlpha();
    }

    boolean isStateful() {
        return mColors != null && mColors.isStateful();
    }

    int getColorForCurrentState(int defaultColor) {
        if (mColors != null) {
            return mColors.getColorForState(mState, defaultColor);
        } else {
            return defaultColor;
        }
    }

    int getColorForCurrentState() {
        if (mColors != null) {
            return getColorForCurrentState(mColors.getDefaultColor());
        } else {
            return Color.TRANSPARENT;
        }
    }

    boolean applyState(@NonNull int[] state) {
        mState = state;

        int colorForState = getColorForCurrentState();
        int oldColor = mPaint.getColor();
        mPaint.setColor(colorForState);
        return mPaint.getColor() != oldColor;
    }
}

File: iconics-kotlin/src/main/java/com/open/iconics/IconicsBrush.kt
package com.open.iconics

import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Paint
import androidx.annotation.IntRange

/**
 * Helper class to control applying state changes to colors and paint. Also for compact providing
 * into [AnimationProcessor][com.mikepenz.iconics.animation.IconicsAnimationProcessor]
 */
class IconicsBrush<T : Paint>(
    /** Will be used for drawing something (icon, background etc.) */
    val paint: T
) {
    private var state: IntArray? = null

    init {
        paint.alpha = 255
    }

    /** Colors which applied on [paint] for drawing current state */
    var colorsList: ColorStateList? = null

    /** Alpha channel for colors */
    var alpha: Int
        get() = paint.alpha
        set(@IntRange(from = 0, to = 255) alpha) {
            if (paint.alpha != alpha) {
                paint.alpha = alpha
            }
        }

    val isStateful: Boolean
        get() = colorsList?.isStateful == true

    val colorForCurrentState: Int
        get() = getColorForCurrentState(colorsList?.defaultColor ?: Color.TRANSPARENT)

    private fun getColorForCurrentState(defaultColor: Int): Int {
        return colorsList?.getColorForState(state, defaultColor) ?: defaultColor
    }

    fun applyState(state: IntArray?): Boolean {
        this.state = state

        val colorForState = colorForCurrentState
        val oldColor = paint.color
        paint.color = colorForState
        return colorForState != oldColor
    }

    override fun toString(): String {
        return "color=#${Integer.toHexString(paint.color)}, state=$state, colorList=$colorsList"
    }
}
